Null check connection in NetworkRTCProvider::createResolver completion handler

achristensen07 opened the pull request:
#### 2aae66e0482238d21b5990f8c4324f39daedeaff
<pre>
Null check connection in NetworkRTCProvider::createResolver completion handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=257975">https://bugs.webkit.org/show_bug.cgi?id=257975</a>
rdar://110456387

Reviewed by Brent Fulgham.

In 264586@main I changed Cocoa&apos;s WebRTC DNS lookup code to use WebCore::resolveDNS which
calls its completion block even if the DNS response comes after the connection has been
destroyed.  Add a null check and don&apos;t send a response through a null connection if this happens.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::createResolver):

Canonical link: <a href="https://commits.webkit.org/265084@main">https://commits.webkit.org/265084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7a17359f7a5eceab4e600b2c55f9c6f1b66280e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10022 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10265 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11427 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12006 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12446 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10747 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11580 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8052 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9150 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12348 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9521 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/8690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/12915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1107 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9292 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
```diff
@@ -306,10 +306,14 @@ void NetworkRTCProvider::createResolver(LibWebRTCResolverIdentifier identifier,
-    WebCore::DNSCompletionHandler completionHandler = [this, identifier](auto&& result) {
+    WebCore::DNSCompletionHandler completionHandler = [connection = m_connection, identifier](auto&& result) {
+        ASSERT(isMainRunLoop());
+        if (!connection)
```
WeakPtr's to the rescue! Should we be making similar checks in other places (e.g., `NetworkRTCProvider::attributedBundleIdetnifierFromPageIdentifier`?

brentfulgham replied:
Or in the completion handler we create in `NetworkRTCProvider::createResolver`?

brentfulgham replied:
LOL. Actually, look at <rdar://110383873> and I think the answer is yes.

achristensen07 replied:
attributedBundleIdetnifierFromPageIdentifier already has a null check.  This is adding a WeakPtr check in NetworkRTCProvider::createResolver.  This fixes rdar://110383873

brentfulgham reviewed (approved):
Looks good, but perhaps we should be checking even more places?

achristensen07 committed:
Null check connection in NetworkRTCProvider::createResolver completion handler
https://bugs.webkit.org/show_bug.cgi?id=257975
rdar://110456387

Reviewed by Brent Fulgham.

In 264586@main I changed Cocoa's WebRTC DNS lookup code to use WebCore::resolveDNS which
calls its completion block even if the DNS response comes after the connection has been
destroyed.  Add a null check and don't send a response through a null connection if this happens.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCProvider.cpp:
(WebKit::NetworkRTCProvider::createResolver):

Canonical link: https://commits.webkit.org/265084@main
